### PR TITLE
Do not process Y velocity in differential drive robots

### PR DESCRIPTION
**Pull request by crdelsey**
---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | #514 |
| Primary OS tested on | Ubuntu |
| Robotic platform tested on | TB 3 in Gazebo |

---

## Description of contribution in a few bullet points

* ROS 1 move base does not pass on the current Y velocity data to diff drive robots. This change matches that behavior.
* This also backs out a less general workaround that was introduced previously.

### CODE REVIEW

File: nav2_dwb_controller/nav_2d_utils/include/nav_2d_utils/odom_subscriber.hpp [hunk 1]
```diff
@@ -78,13 +89,31 @@ class OdomSubscriber
     std::lock_guard<std::mutex> lock(odom_mutex_);
     odom_vel_.header = msg->header;
     odom_vel_.velocity.x = msg->twist.twist.linear.x;
-    odom_vel_.velocity.y = msg->twist.twist.linear.y;
+    if (robot_type_ == HOLONOMIC) {
+      odom_vel_.velocity.y = msg->twist.twist.linear.y;
+    } else {
+      odom_vel_.velocity.y = 0;
```
Comment (SteveMacenski) [inline nav2_dwb_controller/nav_2d_utils/include/nav_2d_utils/odom_subscriber.hpp]: I really think this has to do with an issue in the turtlebot code that 100% needs to be resolved. You should never be receiving odometry from a robot that can only move forward and rotate in the Y axis in reality. If Gazebo is giving you back this information, something is terribly wrong. 

This is masking a much larger issue and I think its more dangerous than helpful for the general user. I have to assume that if there's a Y component even though there shouldn't be, there's some component of the odometry in Y that is actually to be projected into X and these values are therefore screwy somehow. 

Patches for simulations probably don't belong on production hardware code. Maybe its a better choice to fork the gazebo diff drive plugin and force it 0 there or have the turtlebot3 org fix the underlying issues that results in this

Comment (SteveMacenski) [reply nav2_dwb_controller/nav_2d_utils/include/nav_2d_utils/odom_subscriber.hpp]: *oh oh oh oh* I mentioned in another ticket that I'd run into this before, so I racked my brain and remember I posted on ROS Answers, here's the thread: https://answers.ros.org/question/283919/move-base-publishing-y-velocities/

Its TF! Something's not publishing enough so there's interpolations happening that incorrectly assign the transforms (and/or footprint)

**Comment by mkhansen-intel** [comment]
@SteveMacenski - can you submit a fix for this then? We're unable to test changes in simulation reliably without this. I think in the ROS answers you mention it may have to do with amcl publishing transforms with respect to base_footprint? 

If you can submit a fix, we'll cancel this one, otherwise I'm inclined to merge this as a temporary fix until we get a better one.

**Comment by crdelsey** [comment]
@SteveMacenski I'm not entirely convinced this is just a workaround for simulation. Unless you can filter out all Y velocity noise from the odometry data on your real robot, it can cause problems for this algorithm. Movebase in ROS 1 removes the Y velocity data for this reason as well.

With any Y velocity noise, you have two choices:
1. You remove Y velocity from consideration like this PR does, or
2. You have checks that say, if Y < epsilon, replace it with 0.

Without one or the other, DWB is going to fail on a real robot. I could do 2 if you think that is better.
Number 2 makes more sense if you expect your robot to slip sideways, and you want to plan taking that into consideration. Number 1 assumes that the robot can't really slip sideways.

**Comment by crdelsey** [comment]
I should point out that right now, the code base is mixed between the two approaches. Some critics take approach 2 and handle the check for < epsilon themselves. Other places assume approach 1 was taken at a higher level. Each critic that uses 2, has their own epsilon value. Some are configurable, some are not.

File: nav2_dwb_controller/nav_2d_utils/include/nav_2d_utils/odom_subscriber.hpp (hunk 1 above)
Comment (crdelsey) [reply nav2_dwb_controller/nav_2d_utils/include/nav_2d_utils/odom_subscriber.hpp]: I looked into that some. There could be a TF problem. The dimensions of the robot used by Gazebo don't match those used by `robot_state_publisher`. Gazebo uses the SDF model while `robot_state_publisher` uses the URDF model.

But I don't think it is the cause of this problem. I can reproduce with no transforms being published or used in the system. Just Gazebo and teleop are running.

**Comment by SteveMacenski** [comment]
Fair enough if you need something in CI. However I might see a clearer way of not modifying production grade code to patch an issue in Turtlebot3 would be to fork/add the gazebo odometry plugin to navigation2 and then hack that to remove the Y velocities for differential drive robots. I don't think in any situation we should be merging code that hides issues because its convenient for something broken upstream for 1 robot model. That plugin can later be removed once its patched and not modify production code.

 @crdelsey Lets touch on that statement real quick

> Unless you can filter out all Y velocity noise from the odometry data on your real robot...

There should never *be* any Y velocity on your differential drive robot to filter out. If you have 2 wheels and encoders the kinematic equations are `(vr-vl)/2` and `(vr+vl)/2` for angular and forward rotation. There is no kinematic expression for Y velocity. Even in adding an IMU to the fusion, that should still not result in any Y velocity of substance if you're fusing it correctly. In no physical world situation should you be seeing any Y velocities > 0.01 being reported. 

If you're working with an omni or ackermann robot, sure that's right but again you shouldnt be filtering it out, so to speak, its real tangible motion and needs to be accounted for.

**Comment by SteveMacenski** [comment]
This PR is hiding a tangible issue by changing production code and I think that's asking for trouble. I'd rather hack something upstream of this where the issue is probably occurring from being misconfigured and come back another day when I have a ROS2 machine to fully debug. From the post I linked to the 2 ways I got it to work:
- changing to base_link
- making sure the static transforms, republishers, and odometry were coming in fast enough to correctly transform (100hz-ish)
- Changing the launch order of things to make sure things were totally initialized before data was flowing in

**Comment by crdelsey** [comment]
> In no physical world situation should you be seeing any Y velocities > 0.01 being reported.

I agree. But even a noisy 0.01 m/s will cause problems. It could trigger the oscillation critic to fail every trajectory.

Right now, the oscillation critic and others work around this by checking whether the velocity is below a threshold.

The problem with the code as it stands right now is that each critic has its own threshold value. Some are configurable; others are not.

What I'm suggesting in my option 2 above, would be to basically toss this change as is, and replace it with the threshold check the critics are already doing. I can then remove the existing checks in the critics that do the same thing.

Here I'll have a check that says - if Y velocity is below a configurable value (lets say 0.001 by default) then it replaces the value with 0. I could do the same for X and theta to be consistent.

Finally, for this problematic turtlebot, I can change the parameter in the params file to be a much larger value that gets ignored. 

The result would be something you'd want in a real robot I believe. The code base would be a tad simpler and more consistent as well. The turtlebot 3 workaround would be confined to a change in the params file.

**Comment by SteveMacenski** [comment]
So what if you have something like visual odometry being utilized, and due to some slippage of your differential drive robot, or some human physically pushes the robot aside, etc that results in actual tangible Y odometric estimates? This would essentially be throwing out useful information from really modern and used methods for odometry outside of just wheel encoders. We'd just be 0ing out those speeds that are being fed into localization filters so the state of the robot that navigation is now using for local control is not completely different than the state of the motion represented by navigation's planning, environmental model, and general state estimation.

**Comment by crdelsey** [comment]
> This would essentially be throwing out useful information from really modern and used methods for odometry outside of just wheel encoders.

I agree. I think #674 is a better solution for that reason. I initially took this approach because this is what ROS 1 `base_local_planner` does.

> If this issue is purely because of the IMU from gazebo, then sure, lets do it.

I've maybe been a bit sloppy in my terminology. The problem that makes this important is that the velocities provided in the /odom topic when running the TB3 model in Gazebo have a large amount of noise at times.  At other times, they are super accurate. This happens to X and Y, but is only a problem with the Y velocity because DWA has no degree of freedom in Y to compensate.

The /odom topic comes from the `gazebo_diff_drive plugin` however, this plugin doesn't use wheel odometry to generate the /odom velocities. Instead, it pulls the velocities straight from the physics model of Gazebo. For this reason, it is acting more like an IMU in that it samples the current velocities and sends them out.

**Comment by crdelsey** [comment]
Closing this PR. The approach in #674 seems likely a better approach.

**Comment by SteveMacenski** [comment]
So gazebo's diff drive plugin in world reporting frame should be perfect, if you look here:
https://github.com/ros-simulation/gazebo_ros_pkgs/blob/crystal/gazebo_plugins/src/gazebo_ros_diff_drive.cpp#L480

https://github.com/ROBOTIS-GIT/turtlebot3/blob/master/turtlebot3_description/urdf/turtlebot3_burger.gazebo.xacro#L51

If pulls the speed and position information from gazebo directly and applies no noise. What you're getting out should be pretty much perfect. This is actually the main reason in simulation I use the IMU plugin, is to generate actual noise to make my EKF output mimic some uncertainty for other downstream needs.

This indicates to me there's something wrong here with the turtlebot setup OR the transform periods that seems to be the only issue here. 

https://github.com/ROBOTIS-GIT/turtlebot3/blob/ros2/turtlebot3_bringup/launch/turtlebot3_state_publisher.launch.py#L36

I don't see a value set for publisher period, unless something changed in ROS2, there should be a parameter for `publish_frequency` for how fast to republish. If the state publisher doesn't keep up then it will need to interpolate values under the hood which would cause what you're experiencing. For instance if you're publishing robot state at 30hz but odom coming in at 50 (or even 30) there's going to be some offset and guess work. Robot state should be published at ~100hz in my experience, even though defaults for ROS1 were 50.

**Pull request closed without merging**